cmd/snap-confine/snap-device-helper-test.c

```diff
@@ -0,0 +1,195 @@
+	for (size_t i = 0; i < strlen(mod_appname); i++) {
+		if (mod_appname[i] == '.') {
+			mod_appname[i] = '_';
+		}
+	}
+	g_debug("appname modified from %s to %s", appname, mod_appname);
+
+	GError *err = NULL;
+
+	GPtrArray *argv = g_ptr_array_new();
+	g_ptr_array_add(argv, sdh_path);
+	g_ptr_array_add(argv, action);
+	g_ptr_array_add(argv, mod_appname);
+	g_ptr_array_add(argv, devpath);
+	g_ptr_array_add(argv, majmin);
+	g_ptr_array_add(argv, NULL);
+
+	int status = 0;
+
+	gboolean ret = g_spawn_sync(NULL, (gchar **) argv->pdata, NULL, 0,
+				    NULL, NULL, NULL, NULL, &status, &err);
+	g_free(mod_appname);
+	g_free(sdh_path);
+	g_ptr_array_unref(argv);
+
+	if (!ret) {
+		g_debug("failed with: %s", err->message);
+		g_error_free(err);
+		return -2;
+	}
+
+	return WEXITSTATUS(status);
+}
+
+struct sdh_test_data {
+	char *action;
+	char *app;
+	char *file_with_data;
+	char *file_with_no_data;
+};
+
+static void test_sdh_action(gconstpointer test_data)
+{
+	struct sdh_test_data *td = (struct sdh_test_data *)test_data;
+
+	gchar *mock_dir = g_dir_make_tmp(NULL, NULL);
+	gchar *app_dir = g_build_filename(mock_dir, td->app, NULL);
+	gchar *with_data = g_build_filename(mock_dir,
+					    td->app,
+					    td->file_with_data,
+					    NULL);
+	gchar *without_data = g_build_filename(mock_dir,
+					       td->app,
+					       td->file_with_no_data,
+					       NULL);
+	int ret = 0;
+	gchar *data = NULL;
+
+	g_assert(g_mkdir_with_parents(app_dir, 0755) == 0);
+	g_free(app_dir);
+
+	g_debug("mock cgroup dir: %s", mock_dir);
+
+	g_setenv("DEVICES_CGROUP", mock_dir, TRUE);
+
+	g_test_queue_destroy((GDestroyNotify) my_unsetenv, "DEVICES_CGROUP");
+	g_test_queue_destroy((GDestroyNotify) rm_rf_tmp_free, mock_dir);
+
+	ret =
+	    run_sdh(td->action, td->app, "/devices/foo/block/sda/sda4", "8:4");
+	g_assert_cmpint(ret, ==, 0);
+	g_assert_true(g_file_get_contents(with_data, &data, NULL, NULL));
+	g_assert_cmpstr(data, ==, "b 8:4 rwm\n");
+	g_clear_pointer(&data, g_free);
+	g_assert(g_remove(with_data) == 0);
+
+	g_assert_false(g_file_get_contents(without_data, &data, NULL, NULL));
+
+	ret = run_sdh(td->action, td->app, "/devices/foo/tty/ttyS0", "4:64");
+	g_assert_cmpint(ret, ==, 0);
+	g_assert_true(g_file_get_contents(with_data, &data, NULL, NULL));
+	g_assert_cmpstr(data, ==, "c 4:64 rwm\n");
+	g_clear_pointer(&data, g_free);
+	g_assert(g_remove(with_data) == 0);
+
+	g_assert_false(g_file_get_contents(without_data, &data, NULL, NULL));
+
+	g_free(with_data);
+	g_free(without_data);
+}
+
+static void test_sdh_err(void)
+{
+	int ret = 0;
+
+	ret = run_sdh("add", "", "/devices/foo/block/sda/sda4", "8:4");
+	g_assert_cmpint(ret, ==, 1);
+	ret = run_sdh("add", "foo_bar", "", "8:4");
+	g_assert_cmpint(ret, ==, 1);
+	ret = run_sdh("add", "foo_bar", "/devices/foo/block/sda/sda4", "");
+	g_assert_cmpint(ret, ==, 0);
+
+	// mock some stuff so that we can reach the 'action' checks
+	gchar *mock_dir = g_dir_make_tmp(NULL, NULL);
```

Comment:
I think we should check for g_dir_make_tmp error here, otherwise we might be scratching our heads if something further down fails.